database/app/dqlite_other.go

```diff
@@ -0,0 +1,211 @@
+//
+// All App instances in a cluster must be created with the same WithVoters
+// setting.
+//
+// The given value must be an odd number greater than one.
+//
+// The default value is 3.
+func WithVoters(n int) Option {
+	return func() {}
+}
+
+// WithStandBys sets the number of nodes in the cluster that should have the
+// StandBy role.
+//
+// When a new node is added to the cluster or it is started again after a
+// shutdown it will be assigned the StandBy role in case there are already
+// enough online voters, but the current number of stand-bys is below n.
+//
+// Similarly when a node with the StandBy role is shutdown gracefully by
+// calling the Handover() method, it will try to transfer its StandBy role to
+// another non-StandBy node, if one is available.
+//
+// All App instances in a cluster must be created with the same WithStandBys
+// setting.
+//
+// The default value is 3.
+func WithStandBys(n int) Option {
+	return func() {}
+}
+
+// WithRolesAdjustmentFrequency sets the frequency at which the current cluster
+// leader will check if the roles of the various nodes in the cluster matches
+// the desired setup and perform promotions/demotions to adjust the situation
+// if needed.
+//
+// The default is 30 seconds.
+func WithRolesAdjustmentFrequency(frequency time.Duration) Option {
+	return func() {}
+}
+
+// WithLogFunc sets a custom log function.
+func WithLogFunc(log client.LogFunc) Option {
+	return func() {}
+}
+
+// WithFailureDomain sets the node's failure domain.
+//
+// Failure domains are taken into account when deciding which nodes to promote
+// to Voter or StandBy when needed.
+func WithFailureDomain(code uint64) Option {
+	return func() {}
+}
+
+// WithNetworkLatency sets the average one-way network latency.
+func WithNetworkLatency(latency time.Duration) Option {
+	return func() {}
+}
+
+// WithSnapshotParams sets the raft snapshot parameters.
+func WithSnapshotParams(params SnapshotParams) Option {
+	return func() {}
+}
+
+// App is a high-level helper for initializing a typical dqlite-based Go
+// application.
+//
+// It takes care of starting a dqlite node and registering a dqlite Go SQL
+// driver.
+type App struct {
+	dir string
+}
+
+// New creates a new application node.
+func New(dir string, options ...Option) (*App, error) {
+	return &App{dir: dir}, nil
+}
+
+// Ready can be used to wait for a node to complete tasks that
+// are initiated at startup. For example a new node will attempt
+// to join the cluster, a restarted node will check if it should
+// assume some particular role, etc.
+//
+// If this method returns without error it means that those initial
+// tasks have succeeded and follow-up operations like Open() are more
+// likely to succeed quickly.
+func (*App) Ready(ctx context.Context) error {
+	return nil
+}
+
+// Open the dqlite database with the given name
+func (a *App) Open(ctx context.Context, name string) (*sql.DB, error) {
+	path := name
+	if name != ":memory:" {
```

Comment:
This should be a sub-string check. It can be ":memory:?_foreign_keys=1" for example.

Reply:
Well, considering this should be only used in tests, maybe we should just always use memory?

Reply:
Unpredictable behaviour if the test happens to get specific model DBs at some point.